File: utils/file_handler.py
import pandas as pd
import json
import io
import streamlit as st

def load_file(uploaded_file):
    """
    Load data from an uploaded CSV or JSON file into a pandas DataFrame
    
    Parameters:
    uploaded_file (UploadedFile): The file uploaded via Streamlit's file_uploader
    
    Returns:
    pandas.DataFrame: The loaded data
    """
    # Get file type from name
    file_type = uploaded_file.name.split('.')[-1].lower()
    
    # Process based on file type
    if file_type == 'csv':
        try:
            # Try to infer encoding and delimiter
            return pd.read_csv(uploaded_file, encoding='utf-8')
        except UnicodeDecodeError:
            # If utf-8 fails, try other common encodings
            try:
                return pd.read_csv(uploaded_file, encoding='latin1')
            except Exception as e:
                raise Exception(f"Failed to parse CSV file: {str(e)}")
        except Exception as e:
            # Handle other CSV parsing errors
            raise Exception(f"Error loading CSV file: {str(e)}")
    
    elif file_type == 'json':
        try:
            # Load JSON data
            json_data = json.load(uploaded_file)
            
            # Handle different JSON structures
            if isinstance(json_data, list):
                # JSON is a list of objects/records
                return pd.json_normalize(json_data)
            elif isinstance(json_data, dict):
                # Check if it's a nested structure with a data array
                if 'data' in json_data and isinstance(json_data['data'], list):
                    return pd.json_normalize(json_data['data'])
                # Otherwise try to convert the dict directly
                else:
                    return pd.DataFrame([json_data])
            else:
                raise Exception("Unsupported JSON structure. Please provide a JSON array of objects.")
        except Exception as e:
            raise Exception(f"Error loading JSON file: {str(e)}")
    
    else:
        raise Exception(f"Unsupported file type: {file_type}. Please upload a CSV or JSON file.")

File: utils/file_handler.js
const fs = require('fs');
const path = require('path');
const { parse } = require('csv-parse');

/**
 * Load data from a CSV or JSON file into a JavaScript object
 * @param {string} filePath - Path to the uploaded file
 * @returns {Promise<Array>} - Parsed data array
 */
const loadFile = (filePath) => {
  return new Promise((resolve, reject) => {
    const fileExtension = path.extname(filePath).toLowerCase();
    
    if (fileExtension === '.json') {
      try {
        const fileData = fs.readFileSync(filePath, 'utf8');
        const jsonData = JSON.parse(fileData);
        
        // Handle array or object formats
        const data = Array.isArray(jsonData) ? jsonData : [jsonData];
        resolve(data);
      } catch (error) {
        reject(new Error(`Error parsing JSON file: ${error.message}`));
      }
    } else if (fileExtension === '.csv') {
      const records = [];
      
      fs.createReadStream(filePath)
        .pipe(parse({
          columns: true,
          skip_empty_lines: true,
          trim: true
        }))
        .on('data', (record) => {
          // Convert numeric string values to numbers
          Object.keys(record).forEach(key => {
            if (!isNaN(record[key]) && record[key] !== '') {
              record[key] = Number(record[key]);
            }
          });
          
          records.push(record);
        })
        .on('end', () => {
          resolve(records);
        })
        .on('error', (error) => {
          reject(new Error(`Error parsing CSV file: ${error.message}`));
        });
    } else {
      reject(new Error('Unsupported file format. Please upload a CSV or JSON file.'));
    }
  });
};

/**
 * Save uploaded file to disk
 * @param {Object} file - File object from multer
 * @param {string} userId - User ID for organizing uploads
 * @returns {Promise<string>} - Path to saved file
 */
const saveFile = (file, userId) => {
  return new Promise((resolve, reject) => {
    const uploadDir = path.join(__dirname, '..', 'uploads', userId);
    
    // Create user directory if it doesn't exist
    if (!fs.existsSync(uploadDir)) {
      fs.mkdirSync(uploadDir, { recursive: true });
    }
    
    const filePath = path.join(uploadDir, file.originalname);
    
    // Write file to disk
    fs.writeFile(filePath, file.buffer, (err) => {
      if (err) {
        return reject(new Error(`Failed to save file: ${err.message}`));
      }
      
      resolve(filePath);
    });
  });
};

module.exports = {
  loadFile,
  saveFile
};